trace/trace.go

```diff
@@ -0,0 +1,125 @@
+package trace
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/sunfmin/reflectutils"
+	kitlog "github.com/theplant/appkit/log"
+	"github.com/theplant/prottp"
+	"google.golang.org/grpc"
+)
+
+// UnaryServerInterceptor new a server interceptor
+func UnaryServerInterceptor(log kitlog.Logger, traces []MethodTrace) grpc.UnaryServerInterceptor {
+	for i := range traces {
+		// MethodName: GetArticleById
+		// FullMethod: /theplant.ec.api.blog.BlogService/GetArticleById
+		traces[i].MethodName = fmt.Sprintf("/%s", traces[i].MethodName)
+	}
+
+	var findTrace = func(name string) *MethodTrace {
+		for _, trace := range traces {
+			if strings.HasSuffix(name, trace.MethodName) {
+				return &trace
+			}
+		}
+
+		return nil
+	}
+
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		defer func() {
+			trace := findTrace(info.FullMethod)
+			if trace == nil {
+				return
+			}
+
+			var result []interface{}
+			if req != nil || !reflect.ValueOf(req).IsNil() {
+				result = append(
+					result,
+					extract(req, requestPrefix, trace.RequestFields)...,
+				)
+			} else {
+				result = append(result, requestPrefix, "<nil>")
+			}
+			if resp != nil && !reflect.ValueOf(resp).IsNil() {
+				result = append(
+					result,
+					extract(resp, responsePrefix, trace.ResponseFields)...,
+				)
+			} else {
+				result = append(result, responsePrefix, "<nil>")
```

Comment:
Similarly, when will the response be `nil`?

Reply:
Only able to happened on the test case. any a proto message nil wrap to interface{} will not be `== nil`.

Only for cautious.